src/Sarif.Multitool/PageCommand.cs

```diff
@@ -0,0 +1,160 @@
+                root = builder.Build(() => _fileSystem.OpenRead(inputFilePath));
+
+                _fileSystem.WriteAllText(mapPath, JsonConvert.SerializeObject(root, Formatting.None));
+            }
+
+            w.Stop();
+            Console.WriteLine($"Done in {w.Elapsed.TotalSeconds:n1}s.");
+
+            return root;
+        }
+
+        private void ExtractPage(PageOptions options, JsonMapNode root)
+        {
+            if (options.RunIndex < 0) { throw new ArgumentOutOfRangeException("runIndex"); }
+            if (options.Index < 0) { throw new ArgumentOutOfRangeException("index"); }
+            if (options.Count < 0) { throw new ArgumentOutOfRangeException("count"); }
+
+            Stopwatch w = Stopwatch.StartNew();
+            Console.WriteLine($"Extracting Page [{options.Index}, {options.Index + options.Count}) from \"{options.InputFilePath}\" into \"{options.OutputFilePath}\"...");
+
+            JsonMapNode runs, run, results;
+
+            // Get 'runs' node from map. If log was too small, page using the object model
+            if (root == null || root.Nodes == null || !root.Nodes.TryGetValue("runs", out runs))
+            {
+                PageViaOm(options);
+                return;
+            }
+
+            // Verify RunIndex in range
+            if (options.RunIndex >= runs.Count)
+            {
+                throw new ArgumentOutOfRangeException($"Page requested for RunIndex {options.RunIndex}, but Log had only {runs.Count} runs.");
+            }
+
+            // Get 'results' from map. If log was too small, page using the object model
+            if (!runs.Nodes.TryGetValue(options.RunIndex.ToString(), out run) || !run.Nodes.TryGetValue("results", out results) || results.ArrayStarts == null)
+            {
+                // Log too small; convert via OM
+                PageViaOm(options);
+                return;
+            }
+
+            if (options.Index + options.Count >= results.Count)
+            {
+                throw new ArgumentOutOfRangeException($"Page requested from Result {options.Index} to {options.Index + options.Count}, but Run has only {results.Count} results.");
+            }
+
+            Console.WriteLine($"Run {options.RunIndex} in \"{options.InputFilePath}\" has {results.Count:n0} results.");
+
+            Func<Stream> inputStreamProvider = () => _fileSystem.OpenRead(options.InputFilePath);
+            long firstResultStart = results.FindArrayStart(options.Index, inputStreamProvider);
+            long lastResultEnd = results.FindArrayStart(options.Index + options.Count + 1, inputStreamProvider) - 1;
+
+            // Build the Sarif Log subset
+            long lengthWritten = 0;
+            byte[] buffer = new byte[64 * 1024];
+
+            using (FileStream output = File.Create(options.OutputFilePath))
+            using (FileStream source = File.OpenRead(options.InputFilePath))
```

Comment:
>FileStream [](start = 19, length = 10)

i see, no pretty print. :)

Reply:
Right. We're copying stream subsets, so we'll echo the formatting in the input and omit indenting between boundaries we're writing.